src/serialization/sb3.js

```diff
@@ -791,30 +819,58 @@ const deserializeFields = function (fields) {
     return obj;
 };
 
+const decodeBlock = function (block) {
+    const extendedOpcode = Array.isArray(block) ? primitiveIdMap[block[0]] : block.opcode;
+    const firstSplit = extendedOpcode.indexOf('_');
+    if (firstSplit < 0) {
+        return {
+            extendedOpcode: extendedOpcode,
+            extensionId: '',
+            opcode: extendedOpcode
+        };
+    }
+    return {
+        extendedOpcode: extendedOpcode,
+        extensionId: extendedOpcode.slice(0, firstSplit),
+        opcode: extendedOpcode.slice(firstSplit + 1)
+    };
+};
+
 /**
- * Covnert serialized INPUT and FIELD primitives back to hydrated block templates.
+ * Convert serialized INPUT and FIELD primitives back to hydrated block templates.
  * Should be able to deserialize a format that has already been deserialized.  The only
  * "east" path to adding new targets/code requires going through deserialize, so it should
  * work with pre-parsed deserialized blocks.
  *
+ * @param {!Runtime} runtime Runtime object to load all structures into.
  * @param {object} blocks Serialized SB3 "blocks" property of a target. Will be mutated.
  * @return {object} input is modified and returned
  */
-const deserializeBlocks = function (blocks) {
+const deserializeBlocks = function (runtime, blocks) {
     for (const blockId in blocks) {
-        if (!Object.prototype.hasOwnProperty.call(blocks, blockId)) {
+        if (!blocks.hasOwnProperty(blockId)) {
             continue;
         }
         const block = blocks[blockId];
+        const {extensionId, extendedOpcode} = decodeBlock(block);
+        const extensionInfo = runtime.getExtensionInfo(extensionId);
```

Comment:
Notes from discussion: In the long run it seems likely that we'll need to do a two-pass approach to loading. In the first pass we'll initiate loading each extension that the project uses, then we'll wait for them all to finish loading, then we'll do a second pass to actually load the blocks. We don't need that just yet since internal extensions can load synchronously, but we'll need it once we start loading extensions asynchronously.

Reply:
To clarify part of the note above:

> We don't need that just yet since internal extensions can load synchronously, but we'll need it once we start loading extensions asynchronously.

We don't need that just yet since _the extensions for core block packages get loaded synchronously when the VM is initialized_, but we'll need it for non-core extensions that start having custom serialization.